### ISSUE TITLE
Add "bring window to front" functionality

### ISSUE THREAD

**Issue by mdboom** [matplotlib/matplotlib#596]
It would be nice to have a method on the figure object that would bring its GUI window to the top of stack -- if this is possible -- in a backend-neutral way.

(Suggested by Marshall Perrin on the PyInt mailing list)

**Comment by wigging** [comment]
Please fix `plt.show()` on the Mac to bring figure windows to the front of the screen.  On small screens the editor can take up the whole window which completely hides the plot figures.  Another annoyance on the Mac is that multiple figure windows appear directly on top of each other - hiding the other figures below the top most figure.
![screen](https://cloud.githubusercontent.com/assets/6828967/6403160/646874b2-bddb-11e4-8b3c-3e7b0b9459d9.png)

**Comment by embray** [comment]
@wigging That's a related but separate issue from this one, IMO.

**Comment by krischer** [comment]
@embray and others as well: Is there an another issue about it? I cannot find it.

I strongly believe that `plt.show()` in non-interactive mode should always bring the window to the front. At least on OSX using the `qt4agg` backend it always hides behind other windows which is a big annoyance. Using the `macosx` backend one cannot event alt+tab to it but has to search the window in expose.

Is there any reason this is not done by default? I've been manually patching my matplotlib installations for a long time now...

**Comment by efiring** [comment]
@krischer I think the problem is simply that no one has come up with a PR to implement what we need:

1) An API for bringing the window to the front; and separately for controlling whether it grabs focus.
2) An implementation of this for each GUI backend, using the API of its native toolkit.
3) Addition of default settings for bring-to-front and grab-focus to the rcParams configuration mechanism.

Such a PR would be universally welcomed, I believe.  I suspect that even a PR that started with a single backend would be accepted, if there were reason to believe that most of the other backends could be supported in a similar way.

**Comment by OceanWolf** [comment]
I thought about implementing something like this in MEP27, as it separates out the ``Window`` logic.

I agree that a call to ``window.show()`` should always raise a window to the top, by the definition of the word, if the window hides at the back where no-one can see it, then you haven't "shown" it.

At the moment ``show`` (at least on GTK3) only does something (showing the window and bringing it to the top) on its first invocation, successive calls to show do absolutely nothing.  I see no reason why it shouldn't also bring the window to the top on successive calls (apart from BC).

In the end I decided against doing this in MEP27 as I thought it simplest to work as **just** a refactor, and leave it for afterwards, but if people feel otherwise I would gladly add it as part of MEP27, less work for me later (i.e. touching all the backends).

**Comment by wigging** [comment]
@krischer How are you manually patching `matplotlib` to bring the figure windows to the front?  I agree that this should be a default feature. It's very annoying when I can't find my plot window :angry:.

**Comment by krischer** [comment]
@efiring, @OceanWolf I unfortunately don't know enough about matplotlib's backend system to make a qualified statement but I'll try to have a look in the next weeks if I find some time.

@wigging Very simple and probably stupid. It works for the `qt4agg` backend and I guess the `qt5agg` one as well. I also could not notice any side effects with some simple Qt based GUIs I have.

```diff
diff --git a/lib/matplotlib/backends/backend_qt5.py b/lib/matplotlib/backends/backend_qt5.py
index 8e72a08..fcbce44 100644
--- a/lib/matplotlib/backends/backend_qt5.py
+++ b/lib/matplotlib/backends/backend_qt5.py
@@ -543,6 +543,7 @@ class FigureManagerQT(FigureManagerBase):
 
     def show(self):
         self.window.show()
+        self.window.raise_()
 
     def destroy(self, *args):
         # check for qApp first, as PySide deletes it in its atexit handler
```

**Comment by OceanWolf** [comment]
@krischer yes, I wanted to do exactly that :).  Also check out my MEP (https://github.com/matplotlib/matplotlib/blob/master/doc/devel/MEP/MEP27.rst), https://github.com/matplotlib/matplotlib/pull/4143 to see what I referred to above.

As one of the goals in MEP27 lies in standardising the API, I think this makes for the ideal time to do it :smile:.

**Comment by gsonnenf** [comment]
Yes, please make this feature.
This works for Tk on Mac (but not on qt):
        fig.canvas.manager.window.attributes('-topmost', 1)
        fig.canvas.manager.window.attributes('-topmost', 0)  
This works for Qt4 on Mac (but not on tk):
        fig.canvas.manager.window.activateWindow()
        fig.canvas.manager.window.raise_()    

and I still cant figure out how to bring the window to the top if using the matplotlib.use('MacOS').

**Comment by efiring** [comment]
From my standpoint as a user, I think this is now the single biggest barrier to smooth interactive work with matplotlib.  Having to go hunt for the figure window every single time one makes a plot is a terrible distraction.  If this problem has to be solved piecemeal, one backend and platform at a time, in a messy fashion, so be it.  Better that than no solution at all.  Qt might be a good first target, with Tk a close second priority.

**Comment by efiring** [comment]
My strong preference for default behavior would be for the window to be raised, so that it is fully visible, but *not* activated.  I would rather let the user click it to activate it.  The reason is that if it is raised and activated, it is too easy for it to intercept keystrokes that were intended for another window.

**Comment by efiring** [comment]
An alternative to the `raise_()` approach for Qt, as yet untested:
https://forum.qt.io/topic/6032/bring-window-to-front-raise-show-activatewindow-don-t-work-on-windows/7
http://stackoverflow.com/questions/1925015/pyqt-always-on-top
What I gather from this, and from the comments above, is that we probably need to use methods that talk to the window managers--and that no matter what we do, some window managers might refuse to take the hint.
I may give this a try--but I would be delighted if someone with more expertise in gui toolkits volunteered to pick this up.

**Comment by OceanWolf** [comment]
AFAIK, I added this functionality for GTK3 in MEP27 (sorry I haven't got back to that yet, now I have lots on my plate, including looking for a job).  I think it very safe to do it in MEP27 as the deprecation cycle will give people an opportunity to adapt their workflows if they have work-flows that depend on the current functionality.

**Comment by efiring** [comment]
@OceanWolf You are using `grab_focus()`.  I don't think that's what we want, or at least it's not what I want.  Rather than directing keystrokes immediately to the window, which is what I think `grab_focus()` does, I want the window to be raised, so that it is visible.  I would prefer that it *not* automatically gobble keystrokes, but I suspect arranging that will be possible only with some platform/desktop combinations.

**Comment by OceanWolf** [comment]
Erm, I think you look at an unrelated method I added to ``FigureCanvasGTK3``...

You need to look at the ``show`` method of ``WindowGTK`` where I implanted it, which does not call that method in ``FigureCanvasGTK3``

**Comment by efiring** [comment]
Now I see, you are using `present()` in `WindowGtk3.show()`, and that does look like exactly the right thing.  Thank you!   You are using `focus()` which calls `grab_focus()` in the `FigureManager.show()`. Maybe that's OK; I'm not sure whether it is a net win or a loss, but I'm skeptical, and from the comments above it looks like I'm not alone in this.  I would omit it.

**Comment by OceanWolf** [comment]
Sorry, I had forgotten that I had called the focus in ``FigureManager.show``, I only thought of ``WindowGTK3.show`` when I wrote that.

AFAIK, we at least need the focusing thing there on the first call to show, to ensure the mpl keystrokes always work on first show, I see that as an issue of standardisation.  On future calls to ``figman.show`` I don't mind.

Focusing canvas without auto-focus on future calls (note performance hit of extra if check as well):
```python
figman.show()
figman.canvas.focus()
```

Not focusing figure with auto-focus:
```python
figman.window.show()
```

**Comment by OceanWolf** [comment]
I strike that if bit out as I have a better way to do that with events, fire a ``window_show`` event from the window, and set it up the callback to autoremove itself from the callback list when the callback gets executed, i.e. a one-shot event :).

**Comment by tacaswell** [comment]
This *should* be fixed.  If you are not seeing this behavior please create a new issue with backend + window enviroment details.

**Comment by stevengj** [comment]
Although new windows are created on top for me with the TkAgg backend, there still doesn't seem to be any backend-independent way to raise an existing figure window to the front, analogous to [Matlab's shg](http://www.mathworks.com/help/matlab/ref/shg.html).

**Comment by WeatherGod** [comment]
I think what you are requesting is subtly different from what is being
discussed in this thread. This was a discussion on how to make the initial
show of a figure come up to the front. What you are asking for is a
programmatic way to bring a particular window to the front in an
interactive context. Could you file a new issue as a feature request? It'll
get more attention that way rather than rezzin' an dead discussion.

**Comment by gsonnenf** [comment]
I'm going to have to disagree with you WeatherGod, this issue was always having a method to programatically bring a window to the front.

As per OP" "It would be nice to have a method on the figure object that would bring its GUI window to the top of stack -- if this is possible -- in a backend-neutral way."

**Comment by efiring** [comment]
@gsonnenf The standard reply is, "A PR would be welcome."  I agree entirely that it would be good to have this capability on all backends and platforms, but I simply don't know how to do it.  One would think such a basic thing would be easy, and all of the GUI toolkits would provide obvious facilities for it, but as far as I can tell this is not the case at all.

**Comment by gsonnenf** [comment]
I had posted a few piecewise solutions previously, if one were to integrate to create a general purpose command that detected which UI toolkit you were using (e.g. if ('Qt4') { bringtotopQt4(); }) and it worked with the popular ones while threw a 'not supported' for the unpopular ones, it would be a good start.

I believe the method to do this on TK and QT4 are already posted on this thread, I'm sure we can find out the other ones, or at least promote adding this feature when we have them documented.

**Comment by stevengj** [comment]
I filed #6378; do you want to continue this discussion there in lieu of re-opening this issue?

**Comment by efiring** [comment]
Yes, this issue is not going to be reopened; the reference to this one in #6378 is adequate for keeping history.

**Comment by Olsthoorn** [comment]
I found this answer for OSX

fig.canvas.manager.window.raise_()

**Comment by Olsthoorn** [comment]
The figure window(s) is/are probably just behind the spyder or editor window. The simplest solution when using a Mac is learning its trackpad gestures (see system preferences/trackpack). "App Exposé" is the one you need. This makes all windows visible, with one move of your fingers on the trackpad. Then click the figure window to raise it. In a python script/program, it works best when you use plt.show() only once at the very end. This ensures that the program is not blocked by plt.show() before finishing, as the program waits until you actually delete the figure. You may print a message just before plt.show() reminding the user and yourself where the plots are. Notice that all figure windows are raised to focus with a single click on the one that's visible. For me this is a convenient solution to the problem.

**Comment by wigging** [comment]
@Olsthoorn I moved away from using `plt.show()` and instead use the following approach which makes the figure window on Macs appear in front of other items on the screen:

```python
plt.ion()
plt.close('all')

plt.figure(1)
plt.plot(x, y)
plt.grid()
```

**Comment by mbrennwa** [comment]
I know that this is an old (and long) thread, but I seems to me like this went all wrong (I hope I am just not understanding things correctly). I actually DO NOT WANT new figure windows to pop to the front. I liked it the way it was before the change (see [here](https://stackoverflow.com/questions/44278369/how-to-keep-matplotlib-python-window-in-background)). So what is really needed is to make this behavior more user-configurable. Is it possible to control the front / background stacking behavior of a matplotlib figure window?

**Comment by efiring** [comment]
It seems to be different for each backend, and also subject to the whims of the desktop environment.  Which backend are you using, and in which environment?

**Comment by mbrennwa** [comment]
I have Ubuntu Linux with Gnome-3 desktop environment, and I use the TkAgg backaend. matplotlib 1.5 behaves as it should, matplotlib 2.0 does not.

Below is a copy of [what I posted on stackexchange](https://stackoverflow.com/questions/44278369/how-to-keep-matplotlib-python-window-in-background):

********************************************************************
I have a python / matplotlib application that frequently updates a plot with new data coming in from a measurement instrument. The plot window should not change from background to foreground (or vice versa) with respect to other windows on my desktop when the plot is updated with new data.

This worked as desired with Python 3 on a machine running Ubuntu 16.10 with matplotlib 1.5.2rc. However, on a different machine with Ubuntu 17.04 and matplotlib 2.0.0, the figure window pops to the front every time the plot is updated with new data.

How can I control the window foreground/background behavior and keep the window focus when updating the plot with new data?

Here's a code example illustrating my plotting routine:

```
import matplotlib
import matplotlib.pyplot as plt
from time import time
from random import random

print ( matplotlib.__version__ )

# set up the figure
fig = plt.figure()
plt.xlabel('Time')
plt.ylabel('Value')
plt.ion()

# plot things while new data is generated:
t0 = time()
t = []
y = []
while True:
    t.append( time()-t0 )
    y.append( random() )
    fig.clear()
    plt.plot( t , y )
    plt.pause(1)
```
********************************************************************

**Comment by efiring** [comment]
Apart from the change suggested in #8692, which won't help you any time soon, I think you can improve your situation by changing your plotting strategy to avoid using `pause`, which is calling `show`, which is raising the window.
Have you tried using the animation module?  See http://matplotlib.org/devdocs/gallery/animation/random_data.html#sphx-glr-gallery-animation-random-data-py for example.  Modifying it to match your example above would involve adding a call to `line.set_xdata`, and setting 'interval' to 1000 (milliseconds).
I think that if you use the animation module you will need to manually put the window behind others once at the start, but then it will stay there.

**Comment by mbrennwa** [comment]
That animation example does the trick of keeping the window in the background (or in the foreground, wherever I put it) even when the data is updated. I could adapt my code example above to use the animation module.

However, I believe the approach in the example will not work in my real-life application, because I cannot leave it to the animation module to ask for new data to update the plot. My application needs direct control to talk to the measurement instrument, and then update the plot as needed. Is there a way to do this?

**Comment by efiring** [comment]
Yes, I think it could be done with the animation module, but figuring out how--via subclassing Animation--might take a bit of time.  There is probably a simpler way, but it is hard to specify it without seeing exactly what you are trying to do.
Suggestions:
1) Put the ion() call at the top, before creating the figure, or run inside of ipython so that you can take advantage of its interactive magic (with no need for the ion call).
2) General good practice: use the object-oriented API, reserving pyplot for a minimum of calls like plt.ion() and plt.subplots().
3) Don't clear the figure unless you need to; if you are making an updating line plot, for example, just use the data updating strategy illustrated in the animation examples.

I think that is about all you will need; no clear, no pause; the redrawing should occur each time you update the data.

**Comment by WeatherGod** [comment]
The animation classes take an object, usually a timer, that emits when to
advance the frame. So, it is possible to replace that timer with your own
class that can do the job. I don't go into details about this particular
use-case, but the animation module is documented pretty thoroughly in my
"Interactive Applications using Matplotlib" book.

**Comment by mbrennwa** [comment]
Ok, I had a look at this.

At first, I thought I'd try your suggestions on how to get this working without the animation module. I sort of have this working. However, it is not possible to resize or move the plot window on the screen. I guess there's some event loop that would need to be called, but I don't know how to fix this. Any suggestions?

Then I thought again about the animation module. I believe I could use this for my purposes if it would just update the plot with the current data from my measurement instrument stored in a buffer. However, I'd need to make the plot window to be "non blocking". In my existing code I use the pyplot.ion() function to get an interactive figure that does not block the code following the pyplot.show(). This is required to execute the code for instrument control and data acquisition. All examples I found for the animation module use a blocking window without the pyplot.ion() function, and I was not able to make the animation work in a non-blocking (interactive) way. Any suggestions?

**Comment by mbrennwa** [comment]
@WeatherGod: thanks for this suggestion. It would be cool if I could read up on this without buying a book. Any suggestions?

**Comment by mbrennwa** [comment]
Ok, I got it working without the animation module. Window resizing and similar actions are taken care of by calling fig.canvas.flush_events() .

Still, it would be nice if the animation module could be used with non-blocking plots. Is this possible? How?

**Comment by danielfaust** [comment]
I'm facing the same issue as @mbrennwa [Stack Overflow question](https://stackoverflow.com/questions/45729092/make-interactive-matplotlib-window-not-pop-to-front-on-each-plot-call-windows-7)

Honestly, there is nothing in the semantics of the word _pause_ which implies that something should be brought to the front. This may be different with the word _show_, but _pause_ definitely should not be doing this.

**Comment by mbrennwa** [comment]
I fully agree with danielfaust.

Here's how I made things work in my application by avoiding the call to the buggy pause(...) command:
https://stackoverflow.com/questions/44278369/how-to-keep-matplotlib-python-window-in-background

**Comment by tacaswell** [comment]
@mbrennwa `FuncAnimation` and friends work just fine with `plt.ion()` (I use it all the time).  What examples are you referring to?

`FuncAnimation` can take a generator as input as well so you can write things like

```python
def get_data():
    while True:
        yield get_some_hardware_data()
    
anim = FuncAnimation(fig, update_func, get_data(), interval=interval)
```

Could @mbrennwa and @danielfaust take a look at https://github.com/matplotlib/matplotlib/pull/4779 which is a start of thoroughly documenting how GUI event loop integration works in the context of Matplotlib?

**Comment by danielfaust** [comment]
@tacaswell I migrated my code over to to `FuncAnimation` when I faced the `plt.pause` issue. But my problem with `FuncAnimation` is that I don't want to use a fixed interval. I want a fetch-data-time + process-data-time + a fixed pause of let's say 5 seconds. I have some variants where the fetching and processing takes more than 5 seconds, others where both together take about 0.1 seconds, so in the case of the 0.1 seconds I'd choose an interval of 1 second, or a pause of 1 second, with the other one also a pause of one second, but I'd have to choose an interval of about 6-10 seconds with `FuncAnimation`. It's far better for me to just add a pause.

**Comment by tacaswell** [comment]
If you have vary computation time on the data retrieval I would create a thread for that work and have it drop data into a queue, something like

```python
import matplotlib.pyplot as plt
from matplotlib.animation import FuncAnimation
import threading
from queue import Queue, Empty
import time
import numpy as np

q = Queue()


def data_getting_worker():
    while True:
        time.sleep(np.random.rand() * 5)
        q.put(np.random.randn(15))
        print('put some data')


th_gettr = threading.Thread(target=data_getting_worker)

data = np.random.randn(15)


def data_reciver_gen():
    global data
    while True:
        try:
            # pull data off the queue until we get to the most recent one
            data = q.get(block=False)
            print('got some data')
        except Empty:
            yield data


fig, ax = plt.subplots()

ln, = ax.plot(data)


def updater(d_in):
    print('updating line')
    ln.set_ydata(d_in)
    return [ln]


th_gettr.start()
anim = FuncAnimation(fig, updater, frames=data_reciver_gen(), interval=1000)

```

Which so long as the processing in `data_getting_worker` is not a tight loop in c you will get a reasonably responsive figure.  You can also use the same pattern with multiprocess to push that work off to a sub-process.

**Comment by mbrennwa** [comment]
@tacaswell: running the data acquisition (and a lot more) in FuncAnimation (or any other GUI for that matter) would be very awkward in my application. Plotting is just a small side job, and it may not even be possible on some machines that are running without a display system. Wrapping the whole code in a FuncAnimation is therefore certainly not the way to go in my case. I will therefore stick to my current solution, and I might improve the handling of the GUI/plotting event loop a bit following this example: https://stackoverflow.com/questions/45729092/make-interactive-matplotlib-window-not-pop-to-front-on-each-update-windows-7/45734500#45734500
